foxglove_bridge_base/include/foxglove_bridge/websocket_server.hpp

```diff
@@ -131,18 +190,23 @@ class Server final : public ServerInterface {
 
   void setSubscribeHandler(SubscribeUnsubscribeHandler handler) override;
   void setUnsubscribeHandler(SubscribeUnsubscribeHandler handler) override;
+  void setClientAdvertiseHandler(ClientAdvertiseHandler handler) override;
+  void setClientUnadvertiseHandler(ClientUnadvertiseHandler handler) override;
+  void setClientMessageHandler(ClientMessageHandler handler) override;
 
   void sendMessage(ChannelId chanId, uint64_t timestamp, std::string_view data) override;
   void sendMessage(ConnHandle clientHandle, ChannelId chanId, uint64_t timestamp,
                    std::string_view data) override;
 
   std::optional<Tcp::endpoint> localEndpoint() override;
+  std::string remoteEndpointString(ConnHandle clientHandle) override;
 
 private:
   struct ClientInfo {
     std::string name;
     ConnHandle handle;
     std::unordered_map<ChannelId, SubscriptionId> subscriptionsByChannel;
+    std::unordered_set<ClientChannelId> advertisedChannels;
```

Comment:
I think we have to choose another data type here. Every client can choose the client channel ID independent of what other clients have chosen.

Reply:
You’re right. In the ROS 2 node I use a map<ConnHandle, unordered_map<ClientChannelId, Advertisement>>. We probably need the same here